core/trino-main/src/main/java/io/trino/connector/system/MaterializedViewSystemTable.java

```diff
@@ -0,0 +1,109 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.connector.system;
+
+import io.trino.FullConnectorSession;
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.QualifiedObjectName;
+import io.trino.security.AccessControl;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTableMetadata;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.connector.InMemoryRecordSet;
+import io.trino.spi.connector.MaterializedViewFreshness;
+import io.trino.spi.connector.RecordCursor;
+import io.trino.spi.connector.SystemTable;
+import io.trino.spi.predicate.TupleDomain;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.connector.system.jdbc.FilterUtil.tablePrefix;
+import static io.trino.metadata.MetadataListing.getMaterializedViews;
+import static io.trino.metadata.MetadataListing.listCatalogs;
+import static io.trino.spi.connector.SystemTable.Distribution.SINGLE_COORDINATOR;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewSystemTable
+        implements SystemTable
+{
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+    private final MaterializedViewsSystemTableAdapter materializedViewAdapter;
+
+    @Inject
+    public MaterializedViewSystemTable(
+            Metadata metadata,
+            AccessControl accessControl,
+            MaterializedViewsSystemTableAdapter materializedViewAdapter)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.materializedViewAdapter = requireNonNull(materializedViewAdapter, "materializedViewAdapter is null");
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return materializedViewAdapter.getTableMetadata();
+    }
+
+    @Override
+    public RecordCursor cursor(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession connectorSession,
+            TupleDomain<Integer> constraint)
+    {
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+
+        List<MaterializedViewInfo> materializedViews = getAllMaterializedViews(session, constraint);
+        List<Object[]> materializedViewRows = materializedViewAdapter.toTableRows(materializedViews);
+
+        InMemoryRecordSet.Builder displayTable = InMemoryRecordSet.builder(getTableMetadata());
+        materializedViewRows.forEach(displayTable::addRow);
```

Comment:
Instead of constructing `List<MaterializedViewInfo>` and then adding it to `displayTable`, we should pass `displayTable` to `getAllMaterializedViews` and do the addRow there itself.

Reply:
I think that it may be cleaner and easer to read when we separate those steps, otherwise getAllMaterializedViews would have multiple responsibilities (lookup MV metadata in catalog and schema, and then map the entities over). I usually prefer readability > performance unless there could me many MVs.

Reply:
We need metadata listing operations to be efficient because the list of objects can potentially be big and various tools often run lots of metadata listing queries to discover metadata.